### PR TITLE
Optionally pass the Beautiful Soup object to constructors

### PR DESCRIPTION
**Pull request by biolds**
Hi!

Thanks for this nice library! I would like to integrate it into the [Sosse](https://gitlab.com/biolds1/sosse) project, a search engine and crawler I'm writing. In this context, I'm extracting the preview while I already have a Soup object with the web page content. So, here is a small patch that permits passing the Soup object without parsing the page again.

### CODE REVIEW

**Comment by meyt** [comment]
Hi @biolds, Thanks for the PR, nice feature. But needs some changes:
- Add tests
- Use `Union` type annotation, we still support Python<3.10

**Comment by biolds** [comment]
Outch, I did not notice I broke so much stuff :)
I added a test, let me know if you expected something different.

File: linkpreview/linkpreview.py [hunk 1]
```diff
@@ -14,13 +17,21 @@
 
 
 class LinkPreview:
-    def __init__(self, link: Link, parser: str = PARSER):
+    def __init__(self, link: Link, parser: Union[str, None] = PARSER, soup: Union[BeautifulSoup, None] = None):
+        if parser and soup:
```
Comment (meyt) [inline linkpreview/linkpreview.py:21]: @biolds No need to check the arguments twice, just in `PreviewBase` is enough.

File: linkpreview/preview/base.py [hunk 2]
```diff
@@ -7,9 +9,20 @@ class PreviewBase(object):  # pragma: nocover
     Base for all web preview.
     """
 
-    def __init__(self, link: Link, parser: str):
+    def __init__(self, link: Link, parser: Union[str, None], soup: Union[BeautifulSoup, None]):
+        if parser and soup:
+            raise Exception(
+                'Only one of `parser` or `soup` argument must be provided to PreviewBase')
+
+        if not parser and not soup:
```
Comment (meyt) [inline linkpreview/preview/base.py:17]: @biolds This kind of validation is not necessary. bs4 is fine without passing `parser`, and also the developer will receive proper exception if not specifing the `soup`.

File: tests/test_generic.py [hunk 3]
```diff
@@ -110,11 +111,21 @@
 )
 def test_generic(tin, tout):
     link = Link("http://localhost", content=get_sample("generic/%s" % tin))
-    preview = Generic(link, parser="html.parser")
+    preview = Generic(link, "html.parser", None)
     for key in tout.keys():
         assert getattr(preview, key) == tout[key]
 
 
+def test_bs4_object():
```
Comment (meyt) [inline tests/test_generic.py:119]: @biolds Please don't touch other test cases, keep this and add another case for when both `parser` and `soup` are set.

**Comment by biolds** [comment]
It's updated.

**Comment by meyt** [comment]
Thanks @biolds, [v0.11.0](https://pypi.org/project/linkpreview/0.11.0/) released.

**Comment by biolds** [comment]
Thanks!

**Pull request merged**